feat(autocomplete): add contextual skip logic for smart autocomplete triggering

markijbema opened the pull request:
## Summary

Implements intelligent autocomplete skip logic to reduce unnecessary LLM requests and improve user experience by avoiding autocomplete in inappropriate contexts.

## Key Features

- **Language-aware statement terminator detection** - Supports 10+ language families with customizable terminators
- **Skip at end of complete statements** - Avoids autocomplete after `;`, `}`, `)`, etc.
- **Skip during mid-word typing** - Prevents interruptions when typing longer identifiers (>2 chars)
- **Comprehensive test coverage** - 313 test cases covering edge cases and language-specific behavior
- **Centralized configuration** - Easy to add support for new languages

## Implementation Details

### New Module: `contextualSkip.ts`

Provides a single entry point `shouldSkipAutocomplete()` that checks:
1. Whether cursor is at end of a complete statement
2. Whether user is mid-word typing (and word is long enough to skip)

### Language Support

- **C-like languages**: JavaScript, TypeScript, C/C++, C#, Java, Kotlin, Scala, Swift, PHP, Dart, CSS/SCSS/Less, JSON
- **Python**: Custom terminators for Python syntax
- **Ruby**: Including `end` keyword support
- **Go, Rust**: Standard C-like terminators
- **Shell**: Including `fi`, `done`, `esac` keywords
- **SQL**: Semicolon terminators
- **Lisp family**: Clojure, Scheme, Elisp, Racket
- **Markup**: HTML, XML, SVG, Vue, Svelte

## Examples

**Autocomplete skipped (no LLM request):**
```typescript
console.log('foo');<CURSOR>
const x = 5;<CURSOR>
}<CURSOR>
function test() { return true; }<CURSOR>
```

**Autocomplete still triggers:**
```typescript
console.<CURSOR>log();  // mid-line, content after cursor
const x =<CURSOR>       // incomplete assignment
function test() {<CURSOR>  // opening brace, expecting body
con<CURSOR>             // short word (≤2 chars)
```

## Changes

- Add `contextualSkip.ts` module with `shouldSkipAutocomplete()` entry point
- Integrate skip logic into `GhostInlineCompletionProvider`
- Add extensive test suite in `contextualSkip.spec.ts` (313 tests)
- Language-specific configuration for easy expansion

## Testing

All tests pass:
- 313 test cases in `contextualSkip.spec.ts`
- Covers edge cases, language-specific behavior, and boundary conditions
- Tests for both skip and trigger scenarios

Commit:
feat(autocomplete): add contextual skip logic for smart autocomplete triggering

Implements intelligent autocomplete skip logic to reduce unnecessary LLM requests
and improve user experience by avoiding autocomplete in inappropriate contexts.

Key features:
- Language-aware statement terminator detection (supports 10+ language families)
- Skip autocomplete at end of complete statements (after ;, }, ), etc.)
- Skip during mid-word typing (for words longer than 2 characters)
- Comprehensive test coverage with 313 test cases
- Centralized configuration for easy language support expansion

Changes:
- Add contextualSkip.ts module with shouldSkipAutocomplete() entry point
- Integrate skip logic into GhostInlineCompletionProvider
- Add extensive test suite covering edge cases and language-specific behavior

Supported languages: JavaScript/TypeScript, Python, Ruby, Go, Rust, Shell,
SQL, Lisp family, C-like languages, and markup languages.

Commit:
simplify

Commit:
refactor: remove whitespace check for multi-character terminators in contextualSkip

- Removed the /\s$/.test(beforeTerminator) condition from multi-character terminator matching
- Multi-character terminators (fi, done, end) now only match at the start of a line
- Updated tests to reflect the new behavior
- All 49 tests passing

Commit:
refactor: remove multi-character terminator check from contextualSkip

- Removed entire multi-character terminator checking logic (lines 175-183)
- Multi-character terminators like 'fi', 'done', 'end' are no longer checked
- Only single-character terminators (;, }, )) are now validated
- Updated tests to reflect new behavior
- All 49 tests passing

Commit:
remove comment

Commit:
refactor: simplify isMidWordTyping to return boolean

- Changed isMidWordTyping return type from object to boolean
- Moved wordLength > 2 logic inside isMidWordTyping function
- Simplified shouldSkipAutocomplete to directly use boolean result
- All 49 tests passing

Commit:
simplify

Commit:
rename

Commit:
refactor: remove redundant lastCharIsWordChar check in isMidWordTyping

If wordLength >= 1, the regex match guarantees the last char is a word character, making the separate check redundant.

Commit:
refactor: simplify hasContentAfter to single assignment

Combine the firstCharAfter extraction and regex test into a single expression using short-circuit evaluation.

Commit:
simplify

markijbema commented:
Added minor changeset so users see stuff is being done, and so that we can release it

Commit:
add changeset

EamonNerbonne commented on src/services/ghost/classic-auto-complete/contextualSkip.ts:
```diff
@@ -0,0 +1,156 @@
+
+export function shouldSkipAutocomplete(prefix: string, suffix: string, languageId?: string): boolean {
+	return isAtEndOfStatement(prefix, suffix, languageId) || isMidWordTyping(prefix, suffix)
+}
```
This file feels too complex.

markijbema commented on src/services/ghost/classic-auto-complete/contextualSkip.ts:
```diff
@@ -0,0 +1,156 @@
+ * Maps VS Code language IDs to language families.
+ * Computed from LANGUAGE_CONFIGS for efficient lookup.
+ */
+const LANGUAGE_FAMILY_MAP: Record<string, string> = LANGUAGE_CONFIGS.reduce(
```
the idea is that we'll have more data per family later, in an intermediate version i had. However we can make that lookup for now, I just like the administration we have

Commit:
refactor(autocomplete): simplify contextualSkip implementation

- Remove redundant family name field and intermediate lookup map
- Keep family grouping in LANGUAGE_CONFIGS to avoid duplication
- Build LANGUAGE_TERMINATORS directly from configs for O(1) lookup
- Remove redundant fallback in isAtEndOfStatement (handled in getTerminatorsForLanguage)
- Add proper multi-character terminator support using endsWith()
- Fix shell test to correctly expect 'fi' to skip (it's a terminator)

Addresses review comments from @EamonNerbonne

Commit:
simplify comments

EamonNerbonne commented on src/services/ghost/classic-auto-complete/contextualSkip.ts:
```diff
@@ -0,0 +1,140 @@
+	const wordMatch = prefix.match(/([a-zA-Z_][a-zA-Z0-9_]*)$/)
+	const lengthOfWordAtEndOfPrefix = wordMatch ? wordMatch[1].length : 0
+
+	return lengthOfWordAtEndOfPrefix > 2 && !suffixStartsWithWordChar
```
shouldn't this be 
```suggestion
	return lengthOfWordAtEndOfPrefix > 2 && suffixStartsWithWordChar
```

?

(boolean logic and all, maybe I'm missing something obvious)

markijbema replied:
no, but it should be different, thx

Commit:
fix: allow mid-word cache lookups while skipping new LLM requests

Move cache lookup before shouldSkipAutocomplete() check so that
partial typing matches from cache are still returned. The contextual
skip logic now only blocks new LLM requests during mid-word typing,
not cached suggestions.

EamonNerbonne reviewed (approved):
Some sanity check ideas that might be off base; use your judgement!

Commit:
Update src/services/ghost/classic-auto-complete/contextualSkip.ts

Co-authored-by: Eamon Nerbonne <eamon@nerbonne.org>

Commit:
fix(autocomplete): skip autocomplete when suffix starts with word character

Change isMidWordTyping logic from AND to OR condition so that
autocomplete is skipped when either the prefix ends with a long word
OR the suffix starts with a word character. This prevents unwanted
completions when cursor is positioned before existing code.
